ballerina/tests/from_json_test.bal

```diff
@@ -67,6 +67,97 @@ isolated function testSimpleJsonToRecord() returns Error? {
+    ReadonlyUserName name;
+    int age;
+|};
+
+@test:Config
+isolated function testSimpleJsonToRecord3() returns Error? {
+    json user = {id: 4012, name: {firstname: "John", lastname: "Doe"}, age: 27};
+    ReadOnlyUserRecord2 r = check parseAsType(user);
+    test:assertEquals(r, {id: 4012, age: 27, name: {firstname: "John", lastname: "Doe"}});
+}
+
+type UserRecord3 ReadOnlyUserRecord2;
+
+@test:Config
+isolated function testSimpleJsonToRecord4() returns Error? {
+    json user = {id: 4012, name: {firstname: "John", lastname: "Doe"}, age: 27};
+    UserRecord3 r = check parseAsType(user);
+    test:assertEquals(r, {id: 4012, age: 27, name: {firstname: "John", lastname: "Doe"}});
+}
+
+@test:Config
+isolated function testSimpleJsonToRecord5() returns Error? {
+    ReadonlyUserName username = {firstname: "John", lastname: "Doe"};
+    record{string firstname; string lastname;} nameRec = username;
+    json user = {id: 4012, name: nameRec.toJson(), age: 27};
+    ReadOnlyUserRecord2 r = check parseAsType(user);
+    test:assertEquals(r, {id: 4012, age: 27, name: {firstname: "John", lastname: "Doe"}});
+}
+
+type ReadOnlyUsersRecord3 readonly & record {|
+    ReadonlyUserId[] ids;
+    ReadonlyUserName[] names;
+|};
+
+@test:Config
+isolated function testSimpleJsonToRecord6() returns Error? {
+    json user = {ids: [4012, 4013], names: [{firstname: "John", lastname: "Doe"},
+                                            {firstname: "Jane", lastname: "Doe"}], age: 27};
+    ReadOnlyUsersRecord3 r = check parseAsType(user);
+    test:assertEquals(r, {ids: [4012, 4013], names: [{firstname: "John", lastname: "Doe"},
+                                                    {firstname: "Jane", lastname: "Doe"}]});
+}
+
+type ReadOnlyUsersRecord4 readonly & record {|
+    ReadonlyUserId[] & readonly ids;
+    ReadonlyUserName[] & readonly names;
+|};
+
+@test:Config
+isolated function testSimpleJsonToRecord7() returns Error? {
+    json user = {ids: [4012, 4013], names: [{firstname: "John", lastname: "Doe"},
+                                            {firstname: "Jane", lastname: "Doe"}], age: 27};
+    ReadOnlyUsersRecord4 r = check parseAsType(user);
+    test:assertEquals(r, {ids: [4012, 4013], names: [{firstname: "John", lastname: "Doe"},
+                                                    {firstname: "Jane", lastname: "Doe"}]});
+}
+
+type ReadOnlyUsersRecord5 record {|
+    ReadonlyUserId[] & readonly ids;
+    ReadonlyUserName[] & readonly names;
+|};
+
+@test:Config
+isolated function testSimpleJsonToRecord8() returns Error? {
+    json user = {ids: [4012, 4013], names: [{firstname: "John", lastname: "Doe"},
+                                            {firstname: "Jane", lastname: "Doe"}], age: 27};
+    ReadOnlyUsersRecord5 r = check parseAsType(user);
+    test:assertEquals(r, {ids: [4012, 4013], names: [{firstname: "John", lastname: "Doe"},
+                                                    {firstname: "Jane", lastname: "Doe"}]});
```

Comment:
IMO, we need to test the field and/or entire value immutability.

Reply:
Added https://github.com/ballerina-platform/module-ballerina-data.jsondata/pull/45/commits/8f4c00deb5f8429cccc9b188018fa31c2a68d225